plugins/filter/intersect_data_with.py

```diff
@@ -0,0 +1,200 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2022     , DEMAREST Maxime <maxime@indelog.fr>
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+DOCUMENTATION = r'''
+name: intersect_data_with
+short_description: Intersecting an original data structure with modifications
+                   to create a new one.
+version_added: 5.3.0
+author: DEMAREST Maxime (@indelog)
+description:
+  - Starting from two similar data structure, an original provided by I(_input)
+    and modifications provided by I(modifications), creates a new one which is
+    an intersection of these.
+  - This is designed to easily update configurations structures provided by
+    format like JSON, YAML or TOML by ensuring the presence or the absence of
+    some values in keys/items to the result.
+  - Original and modifications must contain items that be assimilable to
+    dictionaries or lists and can themselves contains values that also be
+    assimilable to a dictionary or a list.
+  - Original data acts as basis for the new one.
+  - Modifications are used to add, update or remove keys/items from original
+    data.
+  - The intersection between original and modifications is done by iterating
+    over all keys/items in original and for each which is also exists in
+    modification for a same path in the structure, compares their values.
+  - When the compared values are in both side assimilable to dictionaries, do a
+    recursive intersection.
+positional: expected
+options:
+  _input:
+    description:
+      - A data structure assimilable to a dictionary or a list that acts to
+        basis for the new one.
+      - It must be a similar type of structure that I(modifications) (both are
+        lists or dictionaries).
+      - All of its items that not be updated or remove by the intersection
+        with modification data be present as they are in the result.
+    type: raw
+    required: true
+extends_documentation_fragment: community.general.data_intersection
+seealso:
+  - module: ansible.utils.update_fact
+    description: Do something similar in another way.
+ '''
+
+EXAMPLES = r'''
+- name: Update original data with modifications
+  ansible.builtin.set_fact:
+    result: >
+      {{ original | community.general.intersect_data_with(
+            modifications, present=true) }}
+  vars:
+    original: {A: {AA: '1'}, B: ['2', '3'], C: '4'}
+    modifications: {A: {AB: '9'}, B: ['8', '2'], C: '7'}
+# "result": {"A": {"AA": "1", "AB": "9"}, "B": ["2", "3", "8"], "C": "7"}}
```

Comment:
This is not an intersection, it is more like an union or a merge. I think the filter name should change.

Also the `_data_ ` in the name is, IMO, redundant.